### ISSUE TITLE
correctness of `window_size`

### ISSUE THREAD

**Issue by 152334H** [Dao-AILab/flash-attention#1133]
the README documentation states that

> If window_size != (-1, -1), implements sliding window local attention. Query at position i
> will only attend to keys between [i - window_size[0], i + window_size[1]] inclusive.

this implies that if you have a sequence of length `S`, then the following options should give the same result:
* `causal=True`
* `window_size=(-1, 0)` -- infinite length behind (causal)
* `window_size=(S, 0)` -- S positions behind (also causal)
* `window_size=(S-1,0)` -- since `max(i) == S-1`, `max(i-window_size[0]) == 0`, so this should still be fully causal

```python
import torch
from flash_attn import flash_attn_func

B, S, H, D = 1, 16384, 8, 128

q, k, v = [torch.randn(B, S, H, D, dtype=torch.float16, device="cuda") for _ in range(3)]
y_causal = flash_attn_func(q, k, v, causal=True)  # causal attention
y_ws_neg = flash_attn_func(q, k, v, window_size=(-1, 0))  # Also causal attention
y_ws_S_0 = flash_attn_func(q, k, v, window_size=(S, 0))  # Also causal attention
y_ws_S_1 = flash_attn_func(q, k, v, window_size=(S - 1, 0))  # Also causal attention

print(y_causal.eq(y_ws_neg).all())
print(y_causal.eq(y_ws_S_0).all())
print(y_causal.eq(y_ws_S_1).all())
```

in practice I observe that **the last point is not true**

```python
$ python3 what.py
tensor(True, device='cuda:0')
tensor(True, device='cuda:0')
tensor(False, device='cuda:0')
```

is my understanding of the documentation mistaken? or is something more subtle implementation-wise going on?

**Comment by 152334H** [comment]
what i find very confusing is that `(S-1,0)` DOES appear to be identical to `causal=True` at low sequence lengths:
```python
import torch
from flash_attn import flash_attn_func

B, S, H, D = 1, 2, 1, 4

q, k, v = [torch.randn(B, S, H, D, dtype=torch.float16, device="cuda") for _ in range(3)]
y_causal = flash_attn_func(q, k, v, causal=True)  # causal attention
y_ws_S_1 = flash_attn_func(q, k, v, window_size=(S - 1, 0))  # Also causal attention
y_ws_S_2 = flash_attn_func(q, k, v, window_size=(S - 2, 0))  # this is not causal attention it is no attention diag basically

print(y_causal.eq(y_ws_S_1).all())
print(y_causal.eq(y_ws_S_2).all())

print(y_causal.mean())
print(y_ws_S_2.mean())
```

```python
tensor(True, device='cuda:0')
tensor(False, device='cuda:0')
tensor(-0.6802, device='cuda:0', dtype=torch.float16)
tensor(-0.7744, device='cuda:0', dtype=torch.float16)
```

**Comment by LeoAtlanto** [comment]
Hi @152334H 

I'm recently doing certain experiments with FA and here are some findings which might be helpful.

(1) `window_size=(S-1,0)` should be equal to `causal` mode and below is my result with your code. There might be some precision issue with your software or hardware version. Anyway, you can use `torch.testing.assert_close` to make some comparisons.
<img width="941" alt="image" src="https://github.com/user-attachments/assets/37dc2264-d16e-4405-9399-f2f7c16941f2">

(2) With some other sliding window attention experiments, I found that the exact window size passed to FA should be `window_size-1`. For example, if you want the window size to be 4096, you should pass `window_size=(4095, 0)` to FA.

**Comment by 152334H** [comment]
okay, this is interesting.

I implement the following on 3 different nodes:

```python
$ cat a.py
import torch
from flash_attn import flash_attn_func

B, S, H, D = 1, 16384, 8, 128

q, k, v = [torch.randn(B, S, H, D, dtype=torch.float16, device="cuda") for _ in range(3)]
y_causal = flash_attn_func(q, k, v, causal=True)  # causal attention
y_ws_neg = flash_attn_func(q, k, v, window_size=(-1, 0))  # Also causal attention
y_ws_S_0 = flash_attn_func(q, k, v, window_size=(S, 0))  # Also causal attention
y_ws_S_1 = flash_attn_func(q, k, v, window_size=(S - 1, 0))  # Also causal attention

print(y_causal.eq(y_ws_neg).all())
print(y_causal.eq(y_ws_S_0).all())
print(y_causal.eq(y_ws_S_1).all())
$ md5sum a.py
95e0ec595d1edb38c1308c6651e803a2  a.py
$ pip freeze | grep -e flash-attn  -e ^torch=
flash-attn==2.6.3
torch==2.4.0
```

on,
* NVIDIA RTX 6000 Ada Generation
* NVIDIA GeForce RTX 3090 

I receive
```python
$ python3 a.py
tensor(True, device='cuda:0')
tensor(True, device='cuda:0')
tensor(False, device='cuda:0')
```

on `NVIDIA H100 80GB HBM3` I receive

```python
$ python3 a.py
tensor(True, device='cuda:0')
tensor(True, device='cuda:0')
tensor(True, device='cuda:0')
```

One potential source of contention: I have 535 drivers on the H100, and 550 drivers on the 6000A/3090. What is your system CUDA/driver version? `apt list --installed | grep -e nvidia -e cuda`?

**Comment by LeoAtlanto** [comment]
Mine is CUDA-12.4 with Driver-535.

Even though the results on your RTX6000/RTX3090 are not as expected, the difference might be very small, which you can use torch.testing.assert_close with some tolerance like 5e-3 to figure out. I've not dived that deep into FA, so I'm not quite sure whether it's a bug or just accuracy issue. 

For reference only~

**Comment by tridao** [comment]
As mentioned, in general it's not a good idea to use equality to compare floating points.
```
In [11]: a = torch.randn(10, dtype=torch.bfloat16, device='cuda')

In [12]: torch.equal(a + 0.3 - 0.3, a)
Out[12]: False
```

**Comment by 152334H** [comment]
I understand that much, but I do not know how I could have predicted ahead-of-time that the processes used for
* `causal=True`
* `window_size=(-1,0)`
* `window_size=(S,0)`

would all appear the same, but `window_size=(S-1,0)` different. Is there any general info available about which one of these options truly trigger the same codepaths, and which lead to accidentally-identical output results?

**Comment by tridao** [comment]
There are 2 code paths, one for local and one for causal. There's no guarantee that they produce identical outputs.

**Comment by tridao** [comment]
There's some code to detect that some local window size is equivalent to causal and run the causal path instead (since causal is faster). That's probably causing what you're observing.
https://github.com/Dao-AILab/flash-attention/blob/3669b25206d5938e3cc74a5f7860e31c38af8204/csrc/flash_attn/flash_api.cpp#L630

**Comment by 152334H** [comment]
thank you for the help!
